File: medication-tracker/packages/frontend/src/types/medication.ts
export interface Medication {
  id: string;
  name: string;
  dosage: {
    amount: string;
    unit: string;
    frequency: MedicationFrequency;
    times_per_day: number;
    specific_times: string[];
  };
  schedule: {
    start_date: string;
    end_date?: string;
    reminder_time: number;
    dose_times: string[];
    timezone: string;
  };
  instructions?: string;
  category?: string;
  compliance: number;
  daily_doses_taken: number;
  is_prn: boolean;
  remaining_doses?: number;
  user_id: string;
  status: MedicationStatus;
  reminderEnabled: boolean;
}

export type MedicationFrequency = 
  | 'daily'
  | 'twice-daily'
  | 'weekly'
  | 'monthly'
  | 'as-needed';

export type MedicationStatus = 
  | 'active'
  | 'inactive'
  | 'completed'
  | 'discontinued';

export interface MedicationFormData {
  name: string;
  dosage: {
    amount: string;
    unit: string;
    frequency: MedicationFrequency;
    times_per_day: number;
    specific_times: string[];
  };
  schedule: {
    start_date: Date;
    end_date?: Date;
    reminder_time: number;
    dose_times: string[];
    timezone: string;
  };
  instructions: string;
  category: string;
  is_prn: boolean;
  remaining_doses?: number;
}

export interface AddMedicationProps {
  open: boolean;
  handleClose: () => void;
  onMedicationAdded: (medication: Medication) => void;
}

export interface EditMedicationProps extends Omit<AddMedicationProps, 'onMedicationAdded'> {
  medication: Medication;
  onMedicationUpdated: (medication: Medication) => void;
}

export interface MedicationListProps {
  medications: Medication[];
  onEdit: (id: string) => void;
  onDelete: (id: string) => void;
  onTakeDose?: (id: string) => void;
  showCompliance?: boolean;
}

export interface MedicationFiltersState {
  search: string;
  category: string;
  status: string;
}

export interface MedicationDose {
  id?: string;
  medicationId: string;
  userId: string;
  takenAt: string;
  scheduledFor: string;
  amount: string;
  skipped?: boolean;
  notes?: string;
}

export interface MedicationSchedule {
  userId: string;
  type: 'fixed_time' | 'flexible' | 'as_needed';
  frequency: MedicationFrequency;
  times?: string[];
  intervalHours?: number;
  daysOfWeek?: number[];
  startDate: string;
  endDate?: string;
  timezone: string;
  reminderEnabled: boolean;
  reminderMinutesBefore?: number;
}

export interface MedicationHistory {
  doses: MedicationDose[];
  compliance: number;
  missedDoses: number;
  totalDoses: number;
}

export interface MedicationAlert {
  type: 'missed' | 'upcoming' | 'interaction' | 'refill';
  medicationId: string;
  timestamp: string;
  message: string;
  priority: 'low' | 'medium' | 'high';
  acknowledged: boolean;
}

File: medication-tracker/backend/app/api/schemas/medication.py
from datetime import datetime, time
from typing import List, Optional
from pydantic import BaseModel, Field

class DosageBase(BaseModel):
    amount: float = Field(..., gt=0)
    unit: str
    frequency: str
    times_per_day: int = Field(..., gt=0)
    specific_times: List[str]

class ScheduleBase(BaseModel):
    start_date: datetime
    end_date: Optional[datetime] = None
    reminder_time: int  # minutes before dose
    dose_times: List[str]
    timezone: str = "UTC"

class MedicationBase(BaseModel):
    name: str
    dosage: DosageBase
    schedule: ScheduleBase
    category: Optional[str] = None
    instructions: Optional[str] = None
    is_prn: bool = False
    min_hours_between_doses: Optional[int] = None
    max_daily_doses: Optional[int] = None
    reason_for_taking: Optional[str] = None
    remaining_doses: Optional[int] = None

class MedicationCreate(MedicationBase):
    user_id: int

class MedicationUpdate(MedicationBase):
    pass

class MedicationResponse(MedicationBase):
    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime
    last_taken: Optional[datetime] = None
    daily_doses_taken: int = 0
    daily_doses_reset_at: Optional[datetime] = None

    class Config:
        from_attributes = True
